#1136 New CustomerInvoicePage refactor two

joshxg opened the pull request:
Fixes #1136

## Change summary

- Put modal logic in the reducer
- Put selection logic in the reducer
- Extracted page info columns out, similar to columns/reducer methods. This will inject the `tableState` with a *method* (Require dispatch) to call which will return the page info columns for a page.
- Put `pageObject` in state (transaction for customerInvoice page)
- Adding items logic as reducer methods (including from master list. Requisition add from masterlist might need to change this logic a bit as it needs a store, simple change though.)
- Deleting items logic as reducer methods
- Added a wrapper around `HeaderRow` and `HeaderCell`
- Fixed `topLeftSection` styles


## Testing

*Internal changes, not really testable?*

### Related areas to think about

Note that the reducer can sometimes be called a couple of times for *any* dispatch (from what I have gathered). This is an optimization. Calling the reducer method is supposed to be non-blocking, which means that after a dispatch a render cycle can occur, which may call the reducer again. As the reducer *should* be pure it *shouldn't* matter. So, any reducer method should take care of any side effects it is executing*. I believe reducer being called can be cancelled and also if the diff is the same the result is not used.

*Added the line in `Transaction.addItemsFromMasterList* to for an early exit for this case

This is pretty significant, since we use database side effects in the reducer. I think it will be fine, we could add some middleware to handle the side effects, but could be over kill?

Chris-Petty commented on src/pages/dataTableUtilities/reducer/reducerMethods.js:
```diff
@@ -252,3 +262,124 @@ export const sortData = (state, action) => {
+ * @param {Object} state  The current state
+ * Action: { type: 'closeBasicModal' }
+ */
+export const closeBasicModal = state => ({ ...state, modalIsOpen: false, modalKey: '' });
```
Not sure if it's a bad practice, but could use the empty string for modalKey as equivalent of `false` and remove modalIsOpen all together (would have to use a boolean evaluation on the empty string on the modal `open` prop though).

joshxg replied:
well I don't think so, `''` is falsey and it's javascript!

a530db1774b92255e7f02aca0051ada655fbc9fb

joshxg replied:
forgot one a87c3f9d7da9cfea9a6590536c926f58a39f520c

Chris-Petty commented on src/pages/dataTableUtilities/reducer/reducerMethods.js:
```diff
@@ -252,3 +262,124 @@ export const sortData = (state, action) => {
+
+  return {
+    ...state,
+    data: newData,
```
While it is a bit janky, I think it's ok so long as 
A: The database.write() is synchronous and runs after the newData slice
B: Isn't too slow (though it's only on deleting so not too problematic if there is some delay)

joshxg replied:
I think it's ok for now. Though some thunkers or something would work well. Possibly have a think about implementing our own simple database middleware, like I think you suggested. Would be an easy migration if we decide to go that route, I think

joshxg commented on src/pages/dataTableUtilities/reducer/reducerMethods.js:
```diff
@@ -252,3 +262,124 @@ export const sortData = (state, action) => {
+  });
+
+  const newDataState = new Map();
+  const newData = backingData.slice();
```
yup, live update

joshxg commented on src/pages/dataTableUtilities/reducer/reducerMethods.js:
```diff
@@ -252,3 +262,124 @@ export const sortData = (state, action) => {
+
+  if (!hasSelection) return state;
+
+  const itemsById = Array.from(dataState.keys()).filter(rowKey => dataState.get(rowKey).isSelected);
```
6e52630ddf7469b54ecbe432dede4f857ed59512

joshxg commented on src/pages/dataTableUtilities/reducer/reducerMethods.js:
```diff
@@ -252,3 +262,124 @@ export const sortData = (state, action) => {
+
+  if (pageObject[field] !== value) {
+    database.write(() => {
+      pageObject[field] = value;
```
Just checked it out. It is possible! Do you think more explicit is better? I'm happy to do that. I figured it would come at some point when the logic between different fields starts to differ

Chris-Petty replied:
I think explicit is better, deleting this method would be mean less invitation to skip being explicit

joshxg replied:
b6575d58661aed26dfe5faca59243fbf1d02adf6
1ad3a1b8d3c3a9e0d03a4cb515cccb1a2ffe342a
3caf7f99901b574ee0e6aec3c89de5ac5169258d
0d04575dbdfd38473fb938025763b700251020e2

added `editTheirRef` and `editComment` reducer methods and actions and removed export of `editPageObject`. I think a `editPageObjectSetter` might be a better idea, rather than check if the `pageObject[field]` is a function, and to call rather than assign, but maybe not. Either way I didn't do the change for `editTotalQuantity`, I think it's fine to figure out a bit later, maybe after a couple of pages to make sure we have a broader scope that: two methods for editting fields/using setters or one and how to run the file system for all these reducers methods/actions (and to consider for not just this, but other hooks that might be used throughout the app, aaaand possibly selectors 🤣 ) in #1092 ?

Chris-Petty commented on src/pages/CustomerInvoicePage.js:
```diff
@@ -74,120 +71,47 @@ export const CustomerInvoicePage = ({
+    routeName,
+    {
+      pageObject: transaction,
+      backingData: transaction.items,
```
This also needs to be sorted, as the when deleting items this is used again. Can either be lazy and add sorted here, or you can pass a method for initial state so you can do the sort once then just slice the result for `data`.

joshxg replied:
hmm, after using `sorted` so much recently for vaccines stuff and knowing it doesn't sort in-place, I some how switched to think it was in place..! - just kept `backingData` as unsorted as it doesn't really need to be, but sorted the actual display data

Chris-Petty commented on src/pages/dataTableUtilities/reducer/reducerMethods.js:
```diff
@@ -252,3 +262,160 @@ export const sortData = (state, action) => {
+  });
+
+  const newDataState = new Map();
+  const newData = backingData.slice();
```
Testing I found a problem with this:
![image](https://user-images.githubusercontent.com/7684221/63568098-47c90c00-c5c8-11e9-9768-2650de990b91.png)
Basically need to reapply the sorting? Or splice the the same rows out of `data` instead of slicing again but that's pretty sketchy (e.g. the delete failed on DB, but rows disappear in UI)

joshxg replied:
14a209667c787a312149128358aca51bb6213f12

improved the deletion speed a loooot!

joshxg commented:
Fixed the finalization behaviour with these commits here: 

c6386789961b07b851fdf72dca622c4f7e5be665
ac2089061b525a8cb710ed2862005e00757221d1
17c110ad96a0892fe4ef7b0b20dccd3869affe6a
55da0752305c42848ae4bc67f964eda4061c778a

Also added a change to `usePageReducer` in order to fix the non-pure reducer problem:

623294cf89ed8aae5a24f795fc4efc0ef11d92eb
891155535fdb7e836bd9ce4ee6130cf8b7455d34
7bb388d8fd2e271e7880fa6d52dbcc3dacb55b82
854c917bb27afc010eccd6cf4a0f0489ac6aaa6d
ebb52285fcbf62517d4f7f47bc40625aba7dd395
b75009b378558b844a007c5748fde1be63732017
2dde2be1a03493549a574c4705848e5692affff3

Chris-Petty commented on src/hooks/usePageReducer.js:
```diff
@@ -32,22 +37,27 @@ const usePageReducer = (
   );
 
-  return [state, dispatch, instantDebouncedDispatch, debouncedDispatch];
+  return [state, thunkDispatcher, instantDebouncedDispatch, debouncedDispatch];
```
I want to keep with the trend and call this `thunker` but don't haha

Chris-Petty commented on src/pages/dataTableUtilities/actions.js:
```diff
@@ -60,31 +61,75 @@ export const closeBasicModal = () => ({
-  type: 'deleteItemsById',
-  pageObjectType,
-  object,
+export const addMasterListItems = objectType => (dispatch, state) => {
```
At this point is it worth while js docing the actions too?

joshxg replied:
Yeah definitely - I think the file structure starts to matter more also (`thunk.js`/`action.js`? or `module1Actions.js` etc? Dunno!) .. I think all of these actions/reducers need to be tidied when we have a solid understanding of the groupings etc?

Chris-Petty replied:
Yea, let it mature a bit first 👍

Chris-Petty commented on src/hooks/usePageReducer.js:
```diff
@@ -32,22 +37,27 @@ const usePageReducer = (
-  const debouncedDispatch = useCallback(debounce(dispatch, debounceTimeout), []);
+  const thunkDispatcher = action => {
+    if (typeof action === 'function') action(dispatch, state);
+    dispatch(action);
```
I'd have thought this line would have to be in the `else` clause of the above `if`? Dispatch is receiving a function as an arg otherwise.

joshxg replied:
oopsies d9b0921ca870399557e490400cb446f387f8a742

joshxg commented on src/pages/dataTableUtilities/actions.js:
```diff
@@ -60,31 +61,75 @@ export const closeBasicModal = () => ({
+
+  if (addedItem) dispatch({ type: 'addItem', item: addedItem });
+  else dispatch(closeBasicModal());
+};
```
Ideally, I think you would mock the database calls in unit tests, and then everything else is essentially pure. One of the advantages of using sagas over thunks. Thunks are quite a lot easier/quicker to implement though!

I *think* you should be able to hook `useReducer` up with `react-native-debugger` and `redux-dev-tools`, but maybe not while we are not using `linkingObjects` with realm

joshxg commented on src/pages/CustomerInvoicePage.js:
```diff
@@ -101,15 +102,18 @@ export const CustomerInvoicePage = ({
+  // updated.
+  if (transaction.isFinalised && data.length !== backingData.length) dispatch(refreshData());
+
+  const pageInfoColumns = useMemo(() => pageInfo(pageObject, dispatch), []);
```
🤷‍♀

Chris-Petty commented on src/pages/CustomerInvoicePage.js:
```diff
@@ -74,120 +72,50 @@ export const CustomerInvoicePage = ({
-        return modalStrings.edit_the_invoice_comment;
-      case THEIR_REF_EDIT:
-        return modalStrings.edit_their_reference;
+  const [tableState, dispatch, instantDebouncedDispatch, debouncedDispatch] = usePageReducer(
```
`tableState` should really be `state` now that it has everything and the kitchen sync. Or `pageState`, but state is fine as it usually implies it holds all state for this parent.

joshxg replied:
0e371798027b366c7f70ae9b1271b341d7fb418f

Chris-Petty commented on src/pages/CustomerInvoicePage.js:
```diff
@@ -105,16 +105,16 @@ export const CustomerInvoicePage = ({
     backingData,
   } = tableState;
 
+  const { isFinalised } = transaction;
```
Get halfway?

joshxg replied:
0475969e5f99e9a8c161a742cd4206e8cc030e1b kind of related to this comment
